src/peft/helpers.py

```diff
@@ -0,0 +1,197 @@
+
+                return self.base_model.generate(**kwargs)
+            else:
+                raise NotImplementedError
+    except:
+        self.base_model.prepare_inputs_for_generation = self.base_model_prepare_inputs_for_generation
+        self.base_model._prepare_encoder_decoder_kwargs_for_generation = (
+            self.base_model_prepare_encoder_decoder_kwargs_for_generation
+        )
+        raise
+    else:
+        self.base_model.prepare_inputs_for_generation = self.base_model_prepare_inputs_for_generation
+        self.base_model._prepare_encoder_decoder_kwargs_for_generation = (
+            self.base_model_prepare_encoder_decoder_kwargs_for_generation
+        )
+        return outputs
+
+
+def update_forward_signature(model: PeftModel):
+    """
+    Updates the forward signature of the PeftModel to include parents class signature
+    Args:
+        model (`PeftModel`): Peft model to update the forward signature
+    Example:
+
+    ```python
+    >>> from transformers import  WhisperForConditionalGeneration
+    >>> from peft import  get_peft_model, LoraConfig,  update_forward_signature
+
+    >>> model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny.en")
+    >>> peft_config = LoraConfig(r=8, lora_alpha=32, lora_dropout=0.1, target_modules=["q_proj", "v_proj"])
+
+    >>> peft_model = get_peft_model(model, peft_config)
+    >>> update_forward_signature(peft_model)
+    ```
+    """
+
+    # Only update signature when the current forward signature only has *args and **kwargs
+    current_signature = inspect.signature(model.forward)
+    if (
+        len(current_signature.parameters) == 2
+        and "args" in current_signature.parameters
+        and "kwargs" in current_signature.parameters
+    ):
+        update_wrapper(
+            default_forward, type(model.get_base_model()).forward, assigned=("__doc__", "__name__", "__annotations__")
+        )
+        model.forward = MethodType(default_forward, model)
+
+
+def update_generate_signature(model: PeftModel):
+    """
+    Updates the generate signature of a PeftModel with overriding generate to include parents class signature
+    Args:
+        model (`PeftModel`): Peft model to update the generate signature
+    Example:
+
+    ```python
+    >>> from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
+    >>> from peft import  get_peft_model, LoraConfig, TaskType,  update_generate_signature
+
+    >>> model_name_or_path = "bigscience/mt0-large"
+    >>> tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
+    >>> model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)
+
+    >>> peft_config = LoraConfig(task_type=TaskType.SEQ_2_SEQ_LM, inference_mode=False, r=8, lora_alpha=32, lora_dropout=0.1)
+    >>> peft_model = get_peft_model(model, peft_config)
+    >>> update_generate_signature(peft_model)
+    >>> help(peft_model.generate)
+    ```
+    """
+    if isinstance(model, PeftModelForSeq2SeqLM):
```

Comment:
Similarly to the check for `forward`, should we also check here if the signature only contains `**kwargs` (or `*args, **kwargs`)? That way, we are future proof.

Reply:
Changed to only update generate functions with the following signatures
`def generate(self, *args, **kwargs)` or `def generate(self, **kwargs)`